Make parallel compression optimization code tidier

yzygitzh opened the pull request:
This commit makes #6262's code change tidier and easier to understand by:

1. Wrapping parallel compression initialization and termination into
   common methods;
2. Wrapping BlockRep initialization, push/pop into common methods;
3. Wrapping file size estimation into common methods;
4. Fixing function declarations that use non-const reference;
5. Fixing some uninitialized variables;
6. Fixing first_block data race;
7. Making BlockRep::status check in BlockBasedTableBuilder::Finish only present
if ok();
8. Making assert(ok()) in BlockBasedTableBuilder::CompressAndVerifyBlock only
present in non-parallel compression mode. In parallel compression mode,
compression will abort if status is not OK;
9. Eliminating potential data race caused by BlockBasedTableBuilder::GetStatus()
and BlockBasedTableBuilder::GetIOStatus() by returning status copy instead of
unprotected reference.

yzygitzh commented:
@siying Please take a look, thanks!

siying commented on table/block_based/block_based_table_builder.cc:
```diff
@@ -381,8 +381,12 @@ struct BlockBasedTableBuilder::Rep {
   uint64_t get_offset() { return offset.load(std::memory_order_relaxed); }
   void set_offset(uint64_t o) { offset.store(o, std::memory_order_relaxed); }
 
+  bool ParallelCompressionEnabled() const {
```
Maybe `IsParallelCompressionEnabled()`?

yzygitzh replied:
Rebased.

siying commented:
Can you also rebase?

yzygitzh commented:
> Can you rebase?

Done.

siying commented:
> > Can you rebase?
> 
> Done.

I can't see it: https://github.com/yzygitzh/rocksdb/commits/p2c-refactor

yzygitzh commented:
Is Facebook internal test failure related?

siying commented on table/block_based/block_based_table_builder.cc:
```diff
@@ -684,6 +766,83 @@ struct BlockBasedTableBuilder::ParallelCompressionRep {
+    block_rep->compressed_data->clear();
+    block_rep_pool.push(block_rep);
+
+    if (first_block) {
```
copy&paste a code analysis warning from internal tool:

There may be a Lock Consistency Violation: Read/Write race. Non-private method rocksdb::BlockBasedTableBuilder::ParallelCompressionRep::ReapBlock reads without synchronization from this->first_block. Potentially races with write in method rocksdb::BlockBasedTableBuilder::ParallelCompressionRep::ReapBlock.

siying commented on table/block_based/block_based_table_builder.cc:
```diff
@@ -684,6 +766,83 @@ struct BlockBasedTableBuilder::ParallelCompressionRep {
+  // Emit a block to compression thread
+  void EmitBlock(BlockRep* block_rep) {
+    // Read out first_block here to avoid data race with BGWorkWriteRawBlock
+    bool tmp_first_block = first_block;
```
copy & paste from a code analysis warning:

There may be a Lock Consistency Violation: Read/Write race. Non-private method rocksdb::BlockBasedTableBuilder::ParallelCompressionRep::EmitBlock reads without synchronization from this->first_block. Potentially races with write in method rocksdb::BlockBasedTableBuilder::ParallelCompressionRep::ReapBlock. (error trace: TV107924839)

siying commented on table/block_based/block_based_table_builder.cc:
```diff
@@ -634,41 +638,119 @@ struct BlockBasedTableBuilder::ParallelCompressionRep {
-        first_block(true),
-        finished(false) {
+        file_size_estimator(),
+        first_block(true) {
```
Our internal tool says this function needs to have "explicit" annotation.

yzygitzh commented:
@siying Sorry for so late reply!

Updates are as following:
1) Fixed the data race for `first_block` by passing the info in `BlockRep` struct.
2) Added "explicit" annotation.
3) Made BGWorkWriteRawBlock exit after popping all blocks so that assertions on `BlockRep::status` in BlockBasedTableBuilder::Finish() can now pass.

@jay-zhuang Could you help review the update? Thanks!

BTW, I'm not sure what's "nullable in push() and pop()". Could you guys also copy that error info to me if feasible?

jay-zhuang commented on table/block_based/block_based_table_builder.cc:
```diff
@@ -339,17 +339,21 @@ struct BlockBasedTableBuilder::Rep {
       std::lock_guard<std::mutex> lock(status_mutex);
       return io_status;
     } else {
       return io_status;
```
Here is an error message from our internal tool, is it valid:

> There may be a Lock Consistency Violation: Read/Write race. Non-private method rocksdb::BlockBasedTableBuilder::Rep::GetIOStatus indirectly reads without synchronization from this->io_status.data_loss_. Potentially races with write in method rocksdb::BlockBasedTableBuilder::Rep::SetIOStatus.

jay-zhuang commented on table/block_based/block_based_table_builder.cc:
```diff
@@ -339,17 +339,21 @@ struct BlockBasedTableBuilder::Rep {
+    if (IsParallelCompressionEnabled()) {
       std::lock_guard<std::mutex> lock(status_mutex);
       return status;
     } else {
```
> There may be a Lock Consistency Violation: Read/Write race. Non-private method rocksdb::BlockBasedTableBuilder::Rep::GetStatus indirectly reads without synchronization from this->status.code_. Potentially races with write in method rocksdb::BlockBasedTableBuilder::Rep::SetStatus.

jay-zhuang commented on table/block_based/block_based_table_builder.cc:
```diff
@@ -644,11 +733,93 @@ struct BlockBasedTableBuilder::ParallelCompressionRep {
+                         const Slice* first_key_in_next_block,
+                         BlockBuilder* data_block) {
+    BlockRep* block_rep =
+        PrepareBlockInternal(compression_type, first_key_in_next_block);
```
Seems there's still concurrent read/write issue:
> There may be a Lock Consistency Violation: Read/Write race. Non-private method `rocksdb::BlockBasedTableBuilder::ParallelCompressionRep::EmitBlock` reads without synchronization from `this->is_waiting_first_block`. Potentially races with write in method `rocksdb::BlockBasedTableBuilder::ParallelCompressionRep::EmitBlock.`

yzygitzh replied:
Replaced it with atomic variable.

jay-zhuang commented on table/block_based/block_based_table_builder.cc:
```diff
@@ -644,11 +733,93 @@ struct BlockBasedTableBuilder::ParallelCompressionRep {
+      return;
+    }
+
+    if (is_waiting_first_block) {
```
> There may be a Lock Consistency Violation: Read/Write race. Non-private method `rocksdb::BlockBasedTableBuilder::ParallelCompressionRep::EmitBlock` reads without synchronization from `this->is_waiting_first_block`. Potentially races with write in method `rocksdb::BlockBasedTableBuilder::ParallelCompressionRep::EmitBlock`

jay-zhuang commented on table/block_based/block_based_table_builder.cc:
```diff
@@ -599,41 +604,125 @@ struct BlockBasedTableBuilder::ParallelCompressionRep {
 
-  bool finished;
-
   ParallelCompressionRep(uint32_t parallel_threads)
```
Lint warning: change it to `explicit ParallelCompressionRep(uint32_t parallel_threads)` for Single-argument constructor

yzygitzh replied:
Fixed.

jay-zhuang commented on table/block_based/block_based_table_builder.cc:
```diff
@@ -599,41 +604,125 @@ struct BlockBasedTableBuilder::ParallelCompressionRep {
-  // non-zero compression ratio.
-  bool first_block;
+  // Estimate output file size when parallel compression is enabled. This is
+  // neccesary because compression & flush are no longer synchronized,
```
typo: `necessary`

yzygitzh replied:
Fixed.

jay-zhuang commented on table/block_based/block_based_table_builder.cc:
```diff
@@ -340,58 +337,57 @@ struct BlockBasedTableBuilder::Rep {
+  }
+
+  Status GetStatus() {
+    if (status_ok.load(std::memory_order_relaxed)) {
```
Why using `memory_order_relaxed` here and also all the other places? Maybe add comments for that.

yzygitzh replied:
Added comments. It's due to both write and read of status are protected by status_mutex, so no strict memory order is required to make the status change visible before status_ok changes.

jay-zhuang replied:
sorry, just to confirm, do you mean `status_ok` doesn't have to be strong consistent between threads?
Seems write to `status_ok` is protected by mutex, but not read like here.

yzygitzh replied:
Yes.

`status_ok` indicates the state of `status`, so we only needs to keep modifications before `status_ok` change visible. However, those modifications will always be visible because they're protected by a mutex. So we don't need to require the memory order of `status_ok`.

jay-zhuang replied:
Got it, maybe we should mention it the comments that `we need to keep modifications of status before status_ok`.

yzygitzh replied:
Sure, updated.

yzygitzh replied:
Updated.

jay-zhuang commented on table/block_based/block_based_table_builder.cc:
```diff
@@ -1163,24 +1253,17 @@ void BlockBasedTableBuilder::WriteRawBlock(const Slice& block_contents,
+  ParallelCompressionRep::BlockRep* block_rep = nullptr;
   while (r->pc_rep->write_queue.pop(slot)) {
     slot->Take(block_rep);
     if (!block_rep->status.ok()) {
```
Another warning:
> There may be a Null Dereference: pointer `block_rep` last assigned on line 1257 could be null and is dereferenced at line 1260, column 10. 

There're also a few Linter error for Null Dereference in other places for `block_rep`. Like:
> There's an unchecked dereference of an object that is nullable. Make sure you null-check the object before you call methods on it, pass it as non-nullable parameter, or dereference it. If you are sure the the object cannot be nullptr, make it explicit with CHECK_NOTNULL() or assert(). `block_rep` is nullable.

yzygitzh replied:
Added asserts for this.

yzygitzh replied:
I reproduced the nullable problem on line 1260 with infer tool locally and fixed it. However I was not able to find other Linter errors you mentioned. I added some guards for `block_rep` manually. Hope that would do the fix.
